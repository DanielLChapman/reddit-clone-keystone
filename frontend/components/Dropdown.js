import React, { useState } from 'react';
import PropTypes from 'prop-types'

function Dropdown(props) {
    /*
        let options = ['Hot', 'Best', 'New', 'Top', 'Rising'];
        const [dropValue, setDropValue] = useState('Hot');
        <Dropdown returnFunc={setDropValue} selected={dropValue} options={options} />    
    */
    return (
        <select value={props.selected} onChange={(e) => {
            props.returnFunc(e.target.value);
        }}>
            {
                props.options && props.options.map((x, i) => (
                    <option key={i} value={x}>{x}</option>
                ))
            }
        </select>
    );
}

Dropdown.propTypes = {
    returnFunc: PropTypes.func.isRequired,
    options: PropTypes.array.isRequired,
    selected: PropTypes.string.isRequired,
};
  

export default Dropdown;