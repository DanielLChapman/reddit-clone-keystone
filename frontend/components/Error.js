import styled from 'styled-components';
import React from 'react';

import PropTypes from 'prop-types';

const ErrorStyles = styled.div`
`;

const Error = ({ error }) => {
  if (!error || !error.message) return null;
  if (error.networkError && error.networkError.result && error.networkError.result.errors.length) {
    return error.networkError.result.errors.map((error, i) => (
      <ErrorStyles key={i}>
        <p data-test="graphql-error">
          <strong>Error: </strong>
          {error.message.replace('GraphQL error: ', '')}
        </p>
      </ErrorStyles>
    ));
  }
  return (
    <ErrorStyles>
      <p data-test="graphql-error">
        <strong>Error: </strong>
        {error.message.replace('GraphQL error: ', '')}
      </p>
    </ErrorStyles>
  );
};

Error.defaultProps = {
  error: {},
};

Error.propTypes = {
  error: PropTypes.object,
};

export default Error;
