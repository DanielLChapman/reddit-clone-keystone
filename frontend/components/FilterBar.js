import React, { Component } from 'react';
import { FaHotjar, FaBahai, FaGopuram, FaChartLine} from 'react-icons/fa';



class FilterBar extends Component {
    constructor(props) {
        super(props);
        this.state = {
            li1: true,
            li2: false,
            li3: false,
            li4: false
        }
    }

    toggleClass = (num) => {
        let state = {
            li1: false,
            li2: false,
            li3: false,
            li4: false
        };
        state['li'+num] = true;
        this.setState({...state});
    }

    render() {
        let style = {
            display: "none"
        };
        if (this.state.li3) {
            style= {
                display: "inline-block"
            };
        }
        return (
            <section className="filter-bar"> 
                <h5>Popular Posts</h5>
                <ul>
                    <li className={this.state.li1 ? 'active' : null} onClick={() => this.toggleClass(1)}><FaHotjar /> Hot</li>
                    <li className={this.state.li2 ? 'active' : null} onClick={() => this.toggleClass(2)}><FaBahai /> New</li>
                    <li className={this.state.li3 ? 'active' : null} onClick={() => this.toggleClass(3)}><FaGopuram /> Top</li>
                    <li className="time-selector-for-filter-bar" style={style}>Today</li>
                    <li className={this.state.li4 ? 'active' : null} onClick={() => this.toggleClass(4)}><FaChartLine /> Rising</li>
                </ul>
            </section>
        );
    }
}

export default FilterBar;