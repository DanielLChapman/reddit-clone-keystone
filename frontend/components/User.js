import { gql, useQuery } from '@apollo/client';

export const CURRENT_USER_QUERY = gql`
  query {
    authenticatedItem {
      ... on User {
        id
        username
        email
        name
        posts {
          votes {
            vflag
          }
        }
        comments {
          votes {
            vflag
          }
        }
        moderating {
          id
          slug
          name
        }
        owner {
          id
          slug
          name
        }
        subreddits {
          id
          slug
          name
        }
        postvotes {
          post {
            title
            id
          }
          id
          vflag
        }
        commentvotes {
          comment {
            id
          }
          id
          vflag
        }
    }
  }
  }
`;

export function useUser() {

  const { data } = useQuery(CURRENT_USER_QUERY);
  let test = {...data};

  
  if (test.authenticatedItem) {
    let x = Object.assign({}, test.authenticatedItem);
    //count up post votes
    let postVotes = 0;
    let commentVotes = 0;
    
    x.posts.forEach((q) => {
      q.votes.forEach((y) => {
        y.vflag === 'Upvote' ? postVotes += 1 : postVotes -= 1;
      })
    });
    //count up comment votes
    x.comments.forEach((q) => {
      q.votes.forEach((y) => {
        y.vflag === 'Upvote' ? commentVotes += 1 : commentVotes -= 1;
      })
    });

    x.postVotes = postVotes;
    x.commentVotes = commentVotes;

    return x;
  } else {
    return data?.authenticatedItem;
  }

  
}
