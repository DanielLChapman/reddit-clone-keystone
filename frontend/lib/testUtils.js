import casual from 'casual';

  
// seed it so we get consistent results
casual.seed(777);

const fakeUser = (overrides) => (
    {
    __typename: 'User',
    id: '4234',
    name: casual.name,
    username: casual.name,
    owner: [],
    email: casual.email,
    permissions: ['ADMIN'],
    moderating: [],
    subreddits: [],
    posts: [],
    postvotes: [],
    firstSignin: false,
    comments: [],
    commentvotes:[],
    ...overrides,
});

const fakeSubreddit = (overrides) => (
    {
        __typename: 'Subreddit',
        id: casual.uuid,
        name: casual.string,
        title: casual.string,
        slug: casual.string,
        sidebar: casual.string,
        rules: casual.string,
        description: casual.string,
        status: 'PUBLIC',
        ownder: fakeUser(),
        moderators: [],
        subscriber: [],
        posts: [],
        createdAt: Date.now(),
    }
);

const fakeLinkPost = (overrides) => (
    {
    __typename: 'Post',
    title: casual.sentence,
    user: fakeUser(),
    subreddit: fakeSubreddit(),
    createdAt: Date.now(),
    votes: [],
    post_slug: casual.string,
    link: 'www.google.com',
    type: 'Link',
    comments: [],
    removed: 'False',
    id: casual.uuid,


});

const fakeMediaPost = (overrides) => (
    {
    __typename: 'Post',
    title: casual.sentence,
    user: fakeUser(),
    subreddit: fakeSubreddit(),
    createdAt: Date.now(),
    votes: [],
    post_slug: casual.string,
    link: 'https://cdn.fileinfo.com/img/ss/sm/png_79.jpg',
    type: 'Link',
    comments: [],
    id: casual.uuid,
    removed: 'False'

});

const fakeTextPost = (overrides) => (
    {
    __typename: 'Post',
    title: casual.sentence,
    content: casual.sentence,
    user: fakeUser(),
    subreddit: fakeSubreddit(),
    createdAt: Date.now(),
    votes: [],
    post_slug: casual.string,
    link: '',
    type: 'Text',
    comments: [],
    id: casual.uuid,
    removed: 'False'

});

const fakeComment = (overrides) => (
    {
    __typename: 'Comment',
    content: casual.sentence,
    user: fakeUser(),
    post: fakeTextPost(),
    createdAt: Date.now(),
    votes: [],
    parent: null,
    id: casual.uuid,
    children: [],

});


// Fake LocalStorage
class LocalStorageMock {
    constructor() {
      this.store = {};
    }
  
    clear() {
      this.store = {};
    }
  
    getItem(key) {
      return this.store[key] || null;
    }
  
    setItem(key, value) {
      this.store[key] = value.toString();
    }
  
    removeItem(key) {
      delete this.store[key];
    }
  }



export {

    LocalStorageMock,
    fakeLinkPost,
    fakeComment,
    fakeMediaPost,
    fakeSubreddit,
    fakeTextPost,
    fakeUser,

  };