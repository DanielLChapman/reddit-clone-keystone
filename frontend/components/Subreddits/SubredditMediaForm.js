import { useMutation } from '@apollo/client';
import gql from 'graphql-tag';
import React from 'react';
import useForm from '../../lib/useForm';
import Error from '../Error';
import Form from '../styles/SubredditSubmitForm';
//refetch queries
//import { id_array  } from '../../lib/allSubreddits';


export const SUBREDDIT_MEDIA_POST_MUTATION = gql`
    mutation SUBREDDIT_MEDIA_POST_MUTATION($title: String!, $content: String,  $subreddit_id: ID!, $type: String!, $link: String ) {
        createVariedPost(
            title: $title,
            content: $content,
            subreddit_id: $subreddit_id,
            type: $type,
            link: $link,
        ) {
            id
            content
            post_slug
            title
        }
    }
`;


function sanitizeString(str){
    str = str.replace(/[^a-z0-9áéíóúñü \.,_-]/gim,"");
    return (str.length > 300) ? str.substr(0, 300-1) + '&hellip;' : str;
}

function SubredditMediaForm(props) {
    const { inputs, handleChange, resetForm } = useForm({
        title: '',
        link: ''
      });

    const [createVariedPost, { data, error, loading}] = useMutation(
        SUBREDDIT_MEDIA_POST_MUTATION
    );

    return (
        <div>
            {error && (
                <Error error={error} />
            )}
            {data && (
                <span>Success. <a href={`/r/${props.subreddit_slug}/comments/${data.createVariedPost.id}/${data.createVariedPost.post_slug}`}>Click here to view your new post!</a></span>
            )}
        <Form onSubmit={async (e) => {
            e.preventDefault();

            let typeInput = 'media';
            if (inputs.link.includes('vimeo')) {
                typeInput = 'vimeo';
            } else if (inputs.link.includes('youtube')) {
                typeInput = 'youtube';
            } else {
                typeInput = 'image'
            }
            //NEED TO CREATE A POSTVOTE
            const res = await createVariedPost({
                variables: {
                    title: sanitizeString(inputs.title),
                    subreddit_id: props.id,
                    content: '',
                    link: inputs.link,
                    type: typeInput,
                }
            });

            /*

            Router.push({
                pathname: `/r/${inputs.name.toLowerCase()}`,
              });
*/
        }}>
            <fieldset>
                    <label htmlFor="title">
                        Title <br />
                        <input 
                            required
                            type="text"
                            id="title"
                            name="title"
                            value={inputs.title}
                            onChange={handleChange}
                        />
                    </label>
                    <label htmlFor="link">
                        Link <br />
                        <input 
                            required
                            type="text"
                            id="link"
                            name="link"
                            value={inputs.link}
                            onChange={handleChange}
                        />
                    </label>
                    <button type="submit">Create Link/Media Post</button>
                    
                </fieldset>



        </Form>
        </div>

    );
}

export default SubredditMediaForm;