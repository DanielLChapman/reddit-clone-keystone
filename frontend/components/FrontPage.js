import Link from 'next/link';
import React from 'react';
import { useUser } from './User';

function FrontPage(props) {
    const user = useUser();
    return (
        <div>
            {user && (
                <Link href="/subreddits/create">Create Your Own Subreddit</Link>
            )}
            
        </div>
    );
}

export default FrontPage;