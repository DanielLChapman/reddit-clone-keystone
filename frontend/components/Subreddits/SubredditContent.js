import Link from 'next/link';
import React from 'react';

import PropTypes from 'prop-types';
import SubredditPostForm from './SubredditPostForm';
import SubredditHomePage from './SubredditHomePage';
import SubredditCommentsContainer from './SubredditCommentsContainer';
import SubredditMediaForm from './SubredditMediaForm';
import { faIgloo } from '@fortawesome/free-solid-svg-icons';
import { useUser } from '../User';
import checkUserToSubreddit from '../../lib/checkUser';


function SubredditContent(props) {
    let user = useUser();

    let ownership = checkUserToSubreddit(user, props.subreddit.moderators, props.subreddit.owner);

    switch(props.type) {
        case 'Home': 
            if (props.slug && props.subreddit) {
                return <SubredditHomePage ownership={ownership || false} slug={props.slug} id={props.subreddit.id} subreddit={props.subreddit}/>
            }
            return <span>Something was missing here</span>
        case 'comment':
            if (props.slug && props.subreddit && props.postslug && props.postid) {
                return <SubredditCommentsContainer ownership={ownership || false} slug={props.slug} subreddit={props.subreddit} postslug={props.postslug} postid={props.postid}/>
            }
            return <span>Something was missing here</span>
        case 'permalinkedcomments':
            if (props.slug && props.subreddit && props.postslug && props.postid && props.commentid && props.commentArray) {
                return <SubredditCommentsContainer commentid={props.commentid} commentArray={props.commentArray} ownership={ownership || false} slug={props.slug} subreddit={props.subreddit} postslug={props.postslug} postid={props.postid}/>
            }
            return <span>Something was missing here!</span>
        case 'form':
            if (props.selftext && props.selftext === "true") {
                return (
                    <SubredditPostForm slug={props.slug} id={props.subreddit.id} />
                )
            }
            return (
                <SubredditMediaForm slug={props.slug} id={props.subreddit.id} />
            )
        default: 
            return (
                <span>Something went wrong...</span>
            )
    }
    

}


export default SubredditContent;