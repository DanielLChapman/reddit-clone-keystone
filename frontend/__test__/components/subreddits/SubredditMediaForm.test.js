import { render, screen, waitFor } from '@testing-library/react';
import { MockedProvider } from '@apollo/react-testing';
import SubredditContent from '../../../components/Subreddits/SubredditContent';
import { CURRENT_USER_QUERY } from '../../../components/User';
import { fakeMediaPost, fakeSubreddit, fakeTextPost, fakeUser  } from '../../../lib/testUtils';
import wait from 'waait';
import { act } from 'react-dom/test-utils';
import userEvent from '@testing-library/user-event';

import { SUBREDDIT_MEDIA_POST_MUTATION } from '../../../components/Subreddits/SubredditMediaForm';

const fakeSubredditObj = fakeSubreddit();
const fakePostObj = fakeMediaPost();
const fakeUserObj = fakeUser();



const mocks = [
    {
        request: {
            query: SUBREDDIT_MEDIA_POST_MUTATION,
            variables: {
                title:fakePostObj.title,
                subreddit_id:fakePostObj.subreddit.id,
                content:"",
                link: "https://cdn.fileinfo.com/img/ss/sm/png_79.jpg",
                type: 'image'
            }
        },
        result: {
            data: {
                createVariedPost: {
                    id: fakePostObj.id,
                    content: fakePostObj.content,
                    post_slug: fakePostObj.post_slug,
                    title: fakePostObj.title,
                }
            }
        }

    },
    {
        request: { query: CURRENT_USER_QUERY },
        result: { data: { authenticatedItem: fakeUserObj}}
    }

  ];

describe('It switches subreddit content based on props', () => {
    it ('tries to renders a text form with being logged in', async () => {
        const { container, debug} = render(
            <MockedProvider mocks={mocks} >
                <SubredditContent type="form" id={fakePostObj.subreddit.id} slug={fakePostObj.subreddit.slug} postslug={fakePostObj.post_slug} postid={fakePostObj.id} subreddit={fakePostObj.subreddit}/>
            </MockedProvider>
         )

        await act(async() => {
            await new Promise(resolve => setTimeout(resolve, 0));
        })

        await screen.findByText('Create Link/Media Post');
        expect(container).toMatchSnapshot();
        
    });

    it ('succesfully submits data and gets a desired response', async () => {
        const { container, debug} = render(
            <MockedProvider mocks={mocks} >
                <SubredditContent type="form" id={fakePostObj.subreddit.id}  subreddit={fakePostObj.subreddit}/>
            </MockedProvider>
         )
       

        await screen.findByText('Create Link/Media Post'); 

        await userEvent.type(screen.getByLabelText('Title'), fakePostObj.title);
        await userEvent.type(screen.getByLabelText('Link'), fakePostObj.link);
        await userEvent.click(screen.getByText('Create Link/Media Post'));

        await waitFor(() => wait(0));

        await screen.findByText('Success.'); 
    })


})