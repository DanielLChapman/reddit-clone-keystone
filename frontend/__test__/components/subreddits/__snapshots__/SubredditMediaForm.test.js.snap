// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`It switches subreddit content based on props tries to renders a text form with being logged in 1`] = `
<div>
  <div>
    <form
      class="SubredditSubmitForm__Form-sc-4m25ol-0 dSAjAw"
    >
      <fieldset>
        <label
          for="title"
        >
          Title 
          <br />
          <input
            id="title"
            name="title"
            required=""
            type="text"
            value=""
          />
        </label>
        <label
          for="link"
        >
          Link 
          <br />
          <input
            id="link"
            name="link"
            required=""
            type="text"
            value=""
          />
        </label>
        <button
          type="submit"
        >
          Create Link/Media Post
        </button>
      </fieldset>
    </form>
  </div>
</div>
`;
